lib/vauth/gsasl.c

```diff
@@ -0,0 +1,123 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) 2020, Simon Josefsson, <simon@josefsson.org>, et al.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ * RFC5802 SCRAM-SHA-1 authentication
+ *
+ ***************************************************************************/
+
+#include "curl_setup.h"
+
+#ifdef USE_GSASL
+
+#include <curl/curl.h>
+
+#include "vauth/vauth.h"
+#include "urldata.h"
+#include "sendf.h"
+
+#include <gsasl.h>
+
+/* The last #include files should be: */
+#include "curl_memory.h"
+#include "memdebug.h"
+
+CURLcode
+Curl_auth_create_scram_cf_message(struct Curl_easy *data,
+                                  const char *userp,
+                                  const char *passwdp,
+                                  struct gsasldata *gsasl,
+                                  char **outptr, size_t *outlen)
+{
+  CURLcode result;
+  char *s;
+  size_t slen;
+  int res;
+
+  res = gsasl_init(&gsasl->ctx);
+  if(res != GSASL_OK) {
+    infof(data, "GNU SASL init (%d):%s\n", res, gsasl_strerror(res));
+    return CURLE_REMOTE_ACCESS_DENIED;
+  }
+
+  res = gsasl_client_start(gsasl->ctx, "SCRAM-SHA-1", &gsasl->client);
+  if(res != GSASL_OK) {
+    infof(data, "SCRAM start (%d):%s\n", res, gsasl_strerror(res));
+    return CURLE_REMOTE_ACCESS_DENIED;
+  }
+
+  gsasl_property_set(gsasl->client, GSASL_AUTHID, userp);
+  gsasl_property_set(gsasl->client, GSASL_PASSWORD, passwdp);
```

Comment:
Isn't it odd that these functions can't return error and yet they copy the strings? Don't they malloc the target buffer? If not, don't they then have a maximum size on the input fields?

Reply:
Thanks for review!  This one got me thinking and it took time to sink in.  The design idea (as I recall it) was that having the API fail is not necessary, since if the property was really required it will fail later on with an error code specifying the property was missing.  I think your reaction to the API behaviour is valid, and that proper design is that it should fail with an error code instead of silently continuing.  I'm going to fix this upstream, since I'm doing API/ABI breaks anyway for the next major stable branch.  I'll return to the other concerns once that has been adressed.

Reply:
Upstream fix in https://git.savannah.gnu.org/cgit/gsasl.git/commit/?id=97be94f317e0026ad172d7e59c833e59db8c1ba4 and I've updated the patch to make use of the new feature once it is available.